core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java

```diff
@@ -2168,7 +2167,14 @@ public Expression implement(
         RexCall call,
         List<Expression> translatedOperands) {
       final Expression operand = translatedOperands.get(0);
-      final UnaryExpression e = Expressions.makeUnary(expressionType, operand);
+
+      final Expression e;
+      if (expressionType == ExpressionType.Negate && operand.type == BigDecimal.class) {
+        e = Expressions.call(BuiltInMethod.UNARY_MINUS.method, operand);
```

Comment:
Have you taken a look at the "backup method" convention of the `BinaryImplementor`? Is that also suitable here?

https://github.com/apache/calcite/blob/280642a02a4bcfd1fb9cbe8c5ab672d3619860e7/core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java#L2090

Reply:
BinaryImplementor ? As far as i known, it only be used in binary operator and problem is unary minus. If I understand it incorrectly, please correct me. thanks

Reply:
The class `BinaryImplementor`'s constructor has a parameter `backupMethodName`:
https://github.com/apache/calcite/blob/280642a02a4bcfd1fb9cbe8c5ab672d3619860e7/core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java#L2088-L2093

My suggestion is to change the `UnaryImplementor` to include a similar argument, e.g.
```java
UnaryImplementor(ExpressionType expressionType, String backupMethodName) {
      this.expressionType = expressionType;
      this.backupMethodName = backupMethodName; 
}
```

Then you can specify `BuiltInMethod.UNARY_MINUS.method` as a backup method of the implementor of the operator `UNARY_MINUS`.